akka-http-core/src/main/scala/akka/http/engine/client/ConnectionPool.scala

```diff
@@ -0,0 +1,213 @@
+          case (r, ctx) ⇒ Try(r.withEffectiveUri(securedConnection = false)) -> ctx
+        })
+        val failureSplit = b.add(new FailureSplit[T])
+        val mainFlow = Flow[(HttpRequest, T)]
+          .groupBy { case (req, _) ⇒ req.uri.authority }
+          .map {
+            case (Uri.Authority(host, port, _), hostRequests) ⇒
+              val poolClientFlow = Http().cachedHostConnectionPool[T](HostConnectionPoolSetup(host.toString(), port, setup))
+              hostRequests.via(poolClientFlow)
+          }
+          .flatten(FlattenStrategy.Concat()) // WRONG! FIXME: Needs to be `FlattenStrategy.Merge` instead! (#15089)
+        val merge = b.add(Merge[(Try[HttpResponse], T)](2, OperationAttributes.name("ConnectionPool.merge")))
+
+        map.outlet ~> failureSplit.in
+        failureSplit.out0 ~> mainFlow ~> merge.in(0)
+        failureSplit.out1 ~> merge.in(1)
+
+        FlowShape(map.inlet, merge.out)
+      }
+    }
+  }
+
+  /**
+   * Starts a new host connection pool and constructs the [[ClientFlow]] for it.
+   *
+   * A host connection pool for a given [[HostConnectionPoolSetup]] is a running stream, whose outside interface is
+   * provided by a [[PoolGateway]] actor. This gateway actor receives response demand and incoming requests from zero
+   * or more materialisations of its [[ClientFlow]]. Each [[ClientFlow]] materialisation consists of a [[PoolClient]]
+   * actor implementing a [[org.reactivestreams.Processor]].
+   * When a [[ClientFlow]] is materialized a [[PoolClient]] actor is started, which immediately registers itself with
+   * the [[PoolGateway]] actor. It can then forward response demand and requests to the gateway actor from whom it
+   * receives request demand and responses.
+   * When a [[ClientFlow]] is completed, error-completed or cancelled the flow's [[PoolClient]] actor unregisters itself
+   * from the gateway. The connection pool stream is kept running, as it may be serving other clients now or in the
+   * future.
+   *
+   * A running host connection pool is stopped after the configured idle timeout
+   */
+  def hostPool(setup: HostConnectionPoolSetup)(implicit system: ActorSystem, fm: FlowMaterializer): HostPool = {
+    val shutdownPromise = Promise[Unit]()
+    val poolGateway = startHostConnectionPool(setup, shutdownPromise)
+    val doShutdown: () ⇒ Future[Unit] = { () ⇒
+      if (!shutdownPromise.isCompleted) // optimization: avoid known dead letter (not guaranteed to work)
+        poolGateway ! Shutdown
+      shutdownPromise.future
+    }
+    val flow = Flow[(HttpRequest, Any)] andThenMat { () ⇒
+      val defaultHeaders = prepareDefaultHeaders(setup)
+      val actor = system.actorOf(Props(new PoolClient(poolGateway, defaultHeaders, setup.setup.settings.maxRetries)),
+        poolClientActorName.next())
+      val mat = Http.HostConnectionPool(setup)(doShutdown)
+      ActorProcessor[(HttpRequest, Any), (Try[HttpResponse], Any)](actor) -> mat
+    }
```

Comment:
My feeling is that, once we move to `mapAsync` the need for the `PoolClient` actor would go away anyway.
So I don't yet see where I'd need a "transformation stage that materialized into a value". But I guess I'll find out once I get there.